src/kboot_atc.c

```diff
@@ -60,6 +99,123 @@ static const struct adt_tunable_info atc_tunables[] = {
+    {0x23d2bc430, 26, 3, AUSPLL_DCO_EFUSE_SPARE, AUSPLL_RODCO_BIAS_ADJUST_EFUSE},
+    {0x23d2bc434, 2, 2, AUSPLL_FRACN_CAN, AUSPLL_DLL_START_CAPCODE},
+    {0x23d2bc430, 31, 3, AUSPLL_CLKOUT_DTC_VREG, AUSPLL_DTC_VREG_ADJUST},
+    {0x23d2bc434, 4, 5, AUS_COMMON_SHIM_BLK_VREG, AUS_VREG_TRIM},
+};
+
+static const struct atc_fuse_info atc_fuses_t8103_port1[] = {
+    {0x23d2bc438, 19, 6, CIO3PLL_DCO_NCTRL, CIO3PLL_DCO_COARSEBIN_EFUSE0},
+    {0x23d2bc438, 25, 6, CIO3PLL_DCO_NCTRL, CIO3PLL_DCO_COARSEBIN_EFUSE1},
+    {0x23d2bc438, 31, 1, CIO3PLL_FRACN_CAN, CIO3PLL_DLL_CAL_START_CAPCODE},
+    /* next three rows are some kind of workaround for port 1 */
+    {0x23d2bc438, 14, 5, AUS_COMMON_SHIM_BLK_VREG, AUS_VREG_TRIM},
+    {0x23d2bc43c, 0, 1, CIO3PLL_FRACN_CAN, CIO3PLL_DLL_CAL_START_CAPCODE},
+    {0x23d2bc43c, 1, 3, CIO3PLL_DTC_VREG, CIO3PLL_DTC_VREG_ADJUST},
+    {0x23d2bc438, 7, 2, AUSPLL_DCO_EFUSE_SPARE, AUSPLL_RODCO_ENCAP_EFUSE},
+    {0x23d2bc438, 4, 3, AUSPLL_DCO_EFUSE_SPARE, AUSPLL_RODCO_BIAS_ADJUST_EFUSE},
+    {0x23d2bc438, 12, 2, AUSPLL_FRACN_CAN, AUSPLL_DLL_START_CAPCODE},
+    {0x23d2bc438, 9, 3, AUSPLL_CLKOUT_DTC_VREG, AUSPLL_DTC_VREG_ADJUST},
+    {0x23d2bc438, 14, 4, AUS_COMMON_SHIM_BLK_VREG, AUS_VREG_TRIM},
+};
+
+static const struct atc_fuse_hw atc_fuses[] = {
+    {"atc-phy,t8103", 0, atc_fuses_t8103_port0, ARRAY_SIZE(atc_fuses_t8103_port0)},
+    {"atc-phy,t8103", 1, atc_fuses_t8103_port1, ARRAY_SIZE(atc_fuses_t8103_port1)},
+    {"atc-phy,t6020", -1, NULL, 0},
+};
+
+static u32 read_fuse(const struct atc_fuse_info *fuse)
+{
+    union {
+        u64 dword;
+        u32 words[2];
+        u8 bytes[8];
+    } fuse_data;
+
+    if (fuse->fuse_bit + fuse->fuse_len > 64) {
+        printf("kboot: ATC fuse 0x%lx:%d:%d out of range\n", fuse->fuse_addr, fuse->fuse_bit,
+               fuse->fuse_len);
+        return 0;
+    }
+
+    /* Any other read triggers SErrors */
+    fuse_data.words[0] = read32(fuse->fuse_addr);
+    fuse_data.words[1] = read32(fuse->fuse_addr + 4);
+
+    /*
+     * Assuming we read aa bb cc dd above and have bit_offset 12 and len 4
+     * we have to get rid of aa first to get bb cc dd 00.
+     */
+    const int n_prefix = fuse->fuse_bit / 8;
+    memmove(fuse_data.bytes, fuse_data.bytes + n_prefix, 8 - n_prefix);
+
+    /*
+     * When treating bb cc dd 00 as u32 we get 0x00ddccbb such that we can then
+     * shift it by 4 bits to get 0x000ddccb and then AND it with 0xf to
+     * finally get 0xb which is the value we want.
+     */
+    fuse_data.dword >>= fuse->fuse_bit % 8;
+    fuse_data.dword &= (1ULL << fuse->fuse_len) - 1;
+
+    /*
+     * The additional AND with reg_mask is required because sometimes fuses
+     * contain more bits than the actual register can hold and I'd rather
+     * have the fuse offset/length above match Apple's values than adjusting
+     * them to match the register.
+     */
+    return FIELD_PREP(fuse->reg_mask, fuse_data.dword) & fuse->reg_mask;
```

Comment:
I was expecting that `FIELD_PREP` already ensures that bits outside of the mask can't be set. Care to fix that or create an issue?